lib/de.ml

```diff
@@ -2605,6 +2892,1001 @@ module Def = struct
+      if
+        max_len - len >= 4
+        && unsafe_get_uint32 i (match_pos + len)
+           = unsafe_get_uint32 i (start_pos + len)
+      then lz_extend i start_pos match_pos (len + 4) max_len
+      else _lz_extend i start_pos match_pos len max_len
+
+    let hc_matchfinder_slide_window mf =
+      Array.iteri
+        (fun i v -> mf.hash4_tab.(i) <- v - (window_size land 0xFFFF))
+        mf.hash4_tab
+      ; Array.iteri
+          (fun i v -> mf.next_tab.(i) <- v - (window_size land 0xFFFF))
+          mf.next_tab
+
+    let lz_hash i pos num_bits =
+      let v = unsafe_get_uint32 i pos in
+      Int32.(to_int (shift_right_logical (mul v 0x1E35A7BDl) (32 - num_bits)))
+
+    let rec _matchfinder_longest_rec
+        cur_node best_matchptr os lens mf depth_remaining cutoff =
+      let matchptr = os.i_base + cur_node in
+      if
+        unsafe_get_uint8 os.i (matchptr + lens.best)
+        <> unsafe_get_uint8 os.i (os.i_pos + lens.best)
+      then begin
+        let cur_node = mf.next_tab.(cur_node land (window_size - 1)) in
+        decr depth_remaining
+        ; if cur_node <= cutoff || !depth_remaining = 0 then ()
+          else
+            _matchfinder_longest_rec cur_node best_matchptr os lens mf
+              depth_remaining cutoff
+      end
+      else
+        let len = lz_extend os.i os.i_pos matchptr 0 lens.max in
+        if len >= lens.nice then begin
+          lens.best <- len
+          ; best_matchptr := matchptr
+        end
+        else begin
+          if len > lens.best then begin
+            lens.best <- len
+            ; best_matchptr := matchptr
+          end
+          ; let cur_node = mf.next_tab.(cur_node land (window_size - 1)) in
+            decr depth_remaining
+            ; if cur_node <= cutoff || !depth_remaining = 0 then ()
+              else
+                _matchfinder_longest_rec cur_node best_matchptr os lens mf
+                  depth_remaining cutoff
+        end
+
+    let hc_matchfinder_longest_match mf os lens max_search_depth next_hash =
+      let best_matchptr = ref os.i_pos in
+      let cur_pos = os.i_pos - os.i_base in
+      let depth_remaining = ref max_search_depth in
+      let cur_pos =
+        if cur_pos = window_size then (
+          hc_matchfinder_slide_window mf
+          ; os.i_base <- os.i_base + window_size
+          ; 0)
+        else cur_pos in
+      let cutoff = cur_pos - window_size in
+      if lens.max < 5 then os.i_pos - !best_matchptr
+      else
+        let cur_node = mf.hash4_tab.(!next_hash) in
+        mf.hash4_tab.(!next_hash) <- cur_pos
+        ; mf.next_tab.(cur_pos) <- cur_node
+        ; next_hash := lz_hash os.i (os.i_pos + 1) hc_matchfinder_hash4_order
+        ; if cur_node <= cutoff || lens.best >= lens.nice then
+            os.i_pos - !best_matchptr
+          else begin
+            _matchfinder_longest_rec cur_node best_matchptr os lens mf
+              depth_remaining cutoff
```

Comment:
It seems that `best_matchptr` and `depth_remaining` are references on the heap. According to the `perf` graph:
![perf](https://snipboard.io/v5Hiy7.jpg)

We have a call to the GC and from the assembly generated, a call to `caml_call_gc` is about these values. Can you try to dereference them? At least you can dereference them and re-reference them then in `_matchfind_longest_rec` or try a fully functional implementation.